File: server/src/main/java/dev/slimevr/tracking/trackers/TrackerUtils.java
package dev.slimevr.tracking.trackers;

import java.util.List;


public class TrackerUtils {

	private TrackerUtils() {
	}

	/**
	 * Finds the first tracker from allTrackers matching the position
	 *
	 * @return The tracker as a Tracker
	 */
	public static <T extends Tracker> T findTrackerForBodyPosition(
		List<T> allTrackers,
		TrackerPosition position
	) {
		for (T t : allTrackers) {
			if (t.getBodyPosition() == position)
				return t;
		}
		return null;
	}

	/**
	 * Finds the first non ComputedHumanPoseTracker tracker from allTrackers
	 * matching the position
	 *
	 * @return The non ComputedHumanPoseTracker as a Tracker
	 */
	public static <T extends Tracker> T findNonComputedHumanPoseTrackerForBodyPosition(
		List<T> allTrackers,
		TrackerPosition position
	) {
		for (T t : allTrackers) {
			if (
				t.getBodyPosition() == position
					&& !(t instanceof ComputedHumanPoseTracker)
					&& !(t instanceof HMDTracker)
			)
				return t;
		}
		return null;
	}

	/**
	 * Finds the first non ComputedHumanPoseTracker tracker from allTrackers
	 * matching the position
	 *
	 * @return The non ComputedHumanPoseTracker as a Tracker
	 */
	public static <T extends Tracker> T getHMDTracker(
		List<T> allTrackers
	) {
		for (T t : allTrackers) {
			if (t instanceof HMDTracker && t.getStatus() != TrackerStatus.DISCONNECTED)
				return t;
		}
		return null;
	}

	/**
	 * Returns the first tracker that isn't null out of the 2 trackers passed as
	 * arguments.
	 *
	 * @return The first non-null tracker or null
	 */
	public static <T extends Tracker> T getFirstAvailableTracker(
		T firstTracker,
		T secondTracker
	) {
		if (firstTracker != null)
			return firstTracker;
		return secondTracker;
	}

	/**
	 * Returns the first tracker that isn't null out of the 3 trackers passed as
	 * arguments.
	 *
	 * @return The first non-null tracker or null
	 */
	public static <T extends Tracker> T getFirstAvailableTracker(
		T firstTracker,
		T secondTracker,
		T thirdTracker
	) {
		if (firstTracker != null)
			return firstTracker;
		if (secondTracker != null)
			return secondTracker;
		return thirdTracker;
	}
}

File: server/src/main/java/dev/slimevr/tracking/trackers/TrackerUtils.kt
package dev.slimevr.tracking.trackers

object TrackerUtils {

	/**
	 * Finds the first non-internal tracker from allTrackers
	 * matching the position, that is not DISCONNECTED
	 * These can be IMU (physical/SlimeVR) trackers, trackers from SteamVR (Vive),
	 * or even trackers from third party applications (joycons)
	 *
	 * @return A non-internal tracker
	 */
	@JvmStatic
	fun getNonInternalTrackerForBodyPosition(
		allTrackers: List<Tracker>,
		position: TrackerPosition,
	): Tracker? = allTrackers.firstOrNull {
		it.trackerPosition === position &&
			!it.isInternal &&
			it.status != TrackerStatus.DISCONNECTED &&
			it.status != TrackerStatus.ERROR
	}

	/**
	 * Finds the first non-internal Computed tracker from allTrackers
	 * matching the position, that is not DISCONNECTED.
	 * This should return the HMD or any SteamVR tracker assigned
	 * to the head position.
	 *
	 * @return A non-internal computed tracker (e.g. HMD)
	 */
	@JvmStatic
	fun getNonInternalComputedTrackerForBodyPosition(
		allTrackers: List<Tracker>,
		position: TrackerPosition,
	): Tracker? = allTrackers.firstOrNull {
		it.trackerPosition === position &&
			it.isComputed &&
			!it.isInternal &&
			it.status != TrackerStatus.DISCONNECTED &&
			it.status != TrackerStatus.ERROR
	}

	/**
	 * Returns the first tracker that isn't null out of the 2 trackers passed as
	 * arguments.
	 *
	 * @return The first non-null tracker or null
	 */
	@JvmStatic
	fun getFirstAvailableTracker(
		firstTracker: Tracker?,
		secondTracker: Tracker?,
	): Tracker? = firstTracker ?: secondTracker

	/**
	 * Returns the first tracker that isn't null out of the 3 trackers passed as
	 * arguments.
	 *
	 * @return The first non-null tracker or null
	 */
	@JvmStatic
	fun getFirstAvailableTracker(
		firstTracker: Tracker?,
		secondTracker: Tracker?,
		thirdTracker: Tracker?,
	): Tracker? = firstTracker ?: (secondTracker ?: thirdTracker)
}
